awscli/customizations/assumerole.py

```diff
@@ -36,8 +38,41 @@ def inject_assume_role_provider_cache(session, **kwargs):
                   "assume-role cred provider cache.  Not configuring "
                   "JSONFileCache for assume-role.")
         return
-    provider = cred_chain.get_provider('assume-role')
-    provider.cache = JSONFileCache()
+    cred_chain.get_provider('assume-role').cache = JSONFileCache()
+    saml_provider = cred_chain.get_provider('assume-role-with-saml')
+    saml_provider.cache = JSONFileCache()
+    saml_provider.role_selector = role_selector
+
+
+def role_selector(role_arn, roles):
+    if role_arn:
+        chosen_roles = [r for r in roles if r['RoleArn']==role_arn]
+        if not chosen_roles:
+            raise ValueError(
+                'Your specified role %s does not exist in IdP roles list: %s'
+                % (role_arn, [r['RoleArn'] for r in roles]))
+        role = chosen_roles[0]
+    else:
+        # So end user has not yet configured a role_arn.
+        # If IdP provides more than one role, ask the user to choose one,
+        # otherwise just proceed.
+        if len(roles) > 1:
+            options = [(str(i), role) for i, role in enumerate(roles, start=1)]
+            prompt = "\n".join(
+                "[%s]: %s" % (i, role['RoleArn']) for i, role in options)
+            choice = raw_input(
+                "Please choose the role you would like to assume:\n"
+                "(You can preconfigure the role_arn in your profile)\n"
+                "%s\n"
+                "Selection: " % prompt)
+            role = dict(options).get(choice)
+            if not role:
+                raise ValueError('You selected an invalid role index.')
+        elif roles:
+            role = roles[0]
```

Comment:
Following up from the comment above, I also don't think we should silently pick the first role if only one role is available.  A user doesn't necessarily control which roles they can assume, and something as simple as adding a new role can potentially break them.

There's also no guarantee that the role you can assume could be swapped out by an admin so we'll now silently start assuming some different role the user may be unaware of.  I would expect in this case that we should error out.

I think explicitly requiring the role in the config + `aws configure` helper to make this process easy (by presenting them with a list of roles and asking them to select one) seems like the best way to address this, and cleanly separates concerns.

Does that seem reasonable?